internal/stage2.d: Use processor extensions in @target

the-horo opened the pull request:
With ldc2 -mcpu= or -mattr= can cause llvm internal errors:
```
SplitVectorResult #0: t15: v64i8 = llvm.x86.avx512.pshuf.b.512 TargetConstant:i64<11062>, t12, t13

LLVM ERROR: Do not know how to split the result of this operator!
 #0 0x00007f8cddc05505 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm/17/lib64/libLLVM-17.so+0xa05505)
 #1 0x00007f8cddc02d8f llvm::sys::RunSignalHandlers() (/usr/lib/llvm/17/lib64/libLLVM-17.so+0xa02d8f)
 #2 0x00007f8cddc02f16 (/usr/lib/llvm/17/lib64/libLLVM-17.so+0xa02f16)
 #3 0x00007f8cdc793930 (/usr/lib64/libc.so.6+0x3a930)
 #4 0x00007f8cdc7e4e9c (/usr/lib64/libc.so.6+0x8be9c)
 #5 0x00007f8cdc793886 gsignal (/usr/lib64/libc.so.6+0x3a886)
 #6 0x00007f8cdc77b8b7 abort (/usr/lib64/libc.so.6+0x228b7)
 #7 0x00007f8cdd6d6c8c (/usr/lib/llvm/17/lib64/libLLVM-17.so+0x4d6c8c)
 #8 0x00007f8cddb2f9f5 llvm::report_fatal_error(char const*, bool) (/usr/lib/llvm/17/lib64/libLLVM-17.so+0x92f9f5)
 #9 0x00007f8cde6d27c3 (/usr/lib/llvm/17/lib64/libLLVM-17.so+0x14d27c3)
#10 0x00007f8cde66ded2 (/usr/lib/llvm/17/lib64/libLLVM-17.so+0x146ded2)
#11 0x00007f8cde66ec16 llvm::SelectionDAG::LegalizeTypes() (/usr/lib/llvm/17/lib64/libLLVM-17.so+0x146ec16)
#12 0x00007f8cde7e377a llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/usr/lib/llvm/17/lib64/libLLVM-17.so+0x15e377a)
#13 0x00007f8cde7e8fe3 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/usr/lib/llvm/17/lib64/libLLVM-17.so+0x15e8fe3)
#14 0x00007f8cde7ea7e6 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/usr/lib/llvm/17/lib64/libLLVM-17.so+0x15ea7e6)
#15 0x00007f8ce1c2c089 (/usr/lib/llvm/17/lib64/libLLVM-17.so+0x4a2c089)
#16 0x00007f8cde131951 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/usr/lib/llvm/17/lib64/libLLVM-17.so+0xf31951)
#17 0x00007f8cdddb4d0b llvm::FPPassManager::runOnFunction(llvm::Function&) (/usr/lib/llvm/17/lib64/libLLVM-17.so+0xbb4d0b)
#18 0x00007f8cdddb4f51 llvm::FPPassManager::runOnModule(llvm::Module&) (/usr/lib/llvm/17/lib64/libLLVM-17.so+0xbb4f51)
#19 0x00007f8cdddb59fd llvm::legacy::PassManagerImpl::run(llvm::Module&) (/usr/lib/llvm/17/lib64/libLLVM-17.so+0xbb59fd)
#20 0x000055a87ffc129b (anonymous namespace)::codegenModule(llvm::TargetMachine&, llvm::Module&, char const*, llvm::CodeGenFileType) /mnt/git/ldc/driver/toobj.cpp:146:14
#21 0x000055a87ffc2115 (anonymous namespace)::writeObjectFile(llvm::Module*, char const*) /mnt/git/ldc/driver/toobj.cpp:304:1
#22 0x000055a87ffc33ad writeModule(llvm::Module*, char const*) /mnt/git/ldc/driver/toobj.cpp:499:5
#23 0x000055a87ffbc291 ldc::CodeGenerator::writeAndFreeLLModule(char const*) /mnt/git/ldc/driver/codegenerator.cpp:295:7
#24 0x000055a87ffbbff7 ldc::CodeGenerator::finishLLModule(Module*) /mnt/git/ldc/driver/codegenerator.cpp:259:23
#25 0x000055a87ffbc507 ldc::CodeGenerator::emit(Module*) /mnt/git/ldc/driver/codegenerator.cpp:326:10
#26 0x000055a87ff74b5a codegenModules(Array<Module*>&) /mnt/git/ldc/driver/main.cpp:1304:7
#27 0x000055a87fc74731 mars_mainBody(Param&, Array<char const*>&, Array<char const*>&) /mnt/git/ldc/dmd/main.d:759:9
#28 0x000055a87ff74695 cppmain() /mnt/git/ldc/driver/main.cpp:1243:27
#29 0x000055a87fd905af D main /mnt/git/ldc/driver/main.d:28:1
#30 0x00007f8cdcf6f173 (/usr/lib/gcc/x86_64-pc-linux-gnu/13/libgphobos.so.4+0x56f173)
#31 0x00007f8cdcf6f82d (/usr/lib/gcc/x86_64-pc-linux-gnu/13/libgphobos.so.4+0x56f82d)
#32 0x00007f8cdcf6fa59 _d_run_main2 (/usr/lib/gcc/x86_64-pc-linux-gnu/13/libgphobos.so.4+0x56fa59)
#33 0x00007f8cdcf6fc27 _d_run_main (/usr/lib/gcc/x86_64-pc-linux-gnu/13/libgphobos.so.4+0x56fc27)
#34 0x000055a87ffb0efb args::forwardToDruntime(int, char const**) /mnt/git/ldc/driver/args.cpp:82:1
#35 0x000055a87ff73f3f main /mnt/git/ldc/driver/main.cpp:1123:33
#36 0x00007f8cdc77d2e0 (/usr/lib64/libc.so.6+0x242e0)
#37 0x00007f8cdc77d399 __libc_start_main (/usr/lib64/libc.so.6+0x24399)
#38 0x000055a87f9ae785 _start (/root/ldc-build/bin/ldc2+0x580785)
```

To reproduce this issue is should be enough to do:
```
DFLAGS='-mattr=-avx' dub build --compiler=ldc2
```

A fix seems to only use intrinsics that the target supports rather than using all of them and deciding at runtime which ones to use.

I've changed `sse4.2` to `ssse3` since this is what the intrinsics require., I'm not sure if this is a problem. I've hit this issue initially trying to compile `serve-d` with `ldc2` and `-mcpu=native`.

9il commented on source/mir/ion/internal/stage2.d:
```diff
@@ -197,6 +192,7 @@ version (X86_Any)
     }
 
     version (LDC) @target("arch=skylake-avx512")
+    static if (__traits(targetHasFeature, "avx512bw"))
```
Shouldn't we simply mark this function `version(none)` and remove other changes?

the-horo replied:
I don't think this is the proper fix. The behavior should be that `@target` compiles for a specific architecture with a specific set of features, regardless of the compilation target. This is the behavior seen if `-mattr=-...` is not passed. 

Passing the flag, however, leads to the compilation error I posted above. This is not inherit to the `avx512bw` function, it can be triggered by any of the functions so long as the right `-mattr` is specified on the command line.

Looking at how rust has it implemented, `@target` should make the function be compiled for the given architecture and it should be able to use intrinsics supported by the target. Looking at the source code of ldc2 I've noticed https://github.com/ldc-developers/ldc/blob/master/gen/uda.cpp#L374-L377. If the `-mattr` from the command line overwrites the `@target` from the function it would explain why it would fail to compile, as the only llvm issues I've read that have similar error messages are due to inappropriate target specifications.

Sure enough, removing the `sort` in ldc2 lets the code compile regardless if I put `-mattr=-avx2` or anything else that previously failed. Note that I had to change the `@target` UDA in the `stage2_impl` function to reference the 
CPU extension that is used instead of the architectures, i.e:
```diff
-    version (LDC) @target("arch=broadwell")
+    version (LDC) @target("avx2")
```
but I think this is acceptable as the `if` I've changed can be reverted and the runtime CPU detection can remain.

I can make the appropriate PRs here and at ldc2 in a couple of days, if that's acceptable.

9il replied:
Let's first see what the LDC team thinks about the LDC changes. I am not sure what is the proper fix.

the-horo replied:
I've made the ldc PR and revised the changes in this PR accordingly.

9il replied:
What is the mechanic? For example, does `@target("avx2")` also add `avx` and `sse4.2`?

the-horo replied:
As far as I can tell yes. I've stumbled over https://doc.rust-lang.org/reference/attributes/codegen.html#x86-or-x86_64 which probably applies to ldc2 `@target` as well.

9il replied:
That code is critical for production. I need confirmation from LDC devs or a link to a LLVM codebase.

Please backport the PR for older LDC versions (with `__VERSION__`) when we know which release version will get the required feature. We have to preserve the older behavior for the current compiler versions.

the-horo replied:
x86 target features hierarchy is described at https://llvm.org/doxygen/X86TargetParser_8cpp.html. So, to specifically answer your previous question:
- [avx2](https://llvm.org/doxygen/X86TargetParser_8cpp.html#adfbe529a682e4bb0a963e2d513b35471) enables avx
- [avx](https://llvm.org/doxygen/X86TargetParser_8cpp.html#a5f4642e45dee81e6c8118ab578dac200) enables sse4.2
- sse4.2 -> sse4.1 -> ssse3 -> sse3 -> sse2 -> sse.

The previous archs are also described as a combination of features, for example https://llvm.org/doxygen/X86TargetParser_8cpp.html#a09b67f1241667fcaace45a872b59d4c6

the-horo replied:
There is also the thing that ldc2 doesn't handle `arch=` in `@target` the same way something like clang does. With clang, specifying `target("arch=<cpu>")` will enable the features supported by the cpu, as they appear in the documentation above. ldc2 doesn't do this, it only sets the CPU field in the function: https://github.com/ldc-developers/ldc/blob/908a2e3d816f3e21d82b61d4b0624de92f871feb/gen/uda.cpp#L368-L371. This seems to be enough (i.e. llvm will generate code using instructions available on the target as specified in the attribute) but it fails if one such target feature is explicitly disabled on the command line.

I believe ldc2 should do the same as clang and generate target features according to the cpu passed in `arch=` but this change doesn't seem trivial because the code in clang that handles it is target dependent:
- For x86: https://github.com/llvm/llvm-project/blob/187c035c047616860d31fc399702538ad1f42755/clang/lib/Basic/Targets/X86.cpp#L117-L119
- For aarch64: https://github.com/llvm/llvm-project/blob/187c035c047616860d31fc399702538ad1f42755/clang/lib/Basic/Targets/AArch64.cpp#L1057-L1066 

Since ldc2's handler code is target independent the changes to fully support all targets won't be that small.

Coming back to this PR I still think the changes from `arch=<cpu>` to `<extension>` are needed because calling a function compiled for 1 cpu on another is probably undefined behavior. Limiting the optimizations to a cpu extension such as `avx2` should get all the optimizations that are needed while allowing more portability as any CPU that implements the extension should be able to safely call the function.

Be aware though that I don't much about this topic. I've done my best reading the code and trying to understand what it does but I may have made mistakes. If you are not 100% convinced that what I say is correct do ask someone with more experience since, as you said, you require this code work.

the-horo replied:
I've been mistaken in my previous comment. Even though ldc2 doesn't expand the cpu features they are expanded in llvm at https://llvm.org/doxygen/MCSubtargetInfo_8cpp_source.html#l00175. Since in the llvm source code the cpu features are set before the general target features their value can be overwritten. This is why, when not specifying `-mattr`, `@target("arch=...")` works properly, but when `-mattr` is used it overwrites the cpu features and that generates an error. This is consistent with clang, however, so it's probably the intended behavior.

the-horo replied:
> Please backport the PR for older LDC versions (with `__VERSION__`) when we know which release version will get the required feature. We have to preserve the older behavior for the current compiler versions.

Done. When the ldc version is at least 1.38.0 the code will use the new `@target` attributes.

the-horo committed:
internal/stage2.d: Use processor extensions in `@target`

If compiling with >=ldc-1.38.0 (so that
https://github.com/ldc-developers/ldc/pull/4620 is merged) use
processor extensions in `@target` attribute instead of the cpu name.

Signed-off-by: Andrei Horodniceanu <a.horodniceanu@proton.me>

9il commented on source/mir/ion/internal/stage2.d:
```diff
@@ -196,7 +196,7 @@ version (X86_Any)
     }
 
-    version (LDC) @target("arch=skylake-avx512")
+    version (LDC) @target("avx512bw")
```
Please `version(none)` it because it isn't used anyway
